### PR TITLE
Fix #7311: External configuration file does not change directories

### LINKED ISSUE #7311: Wrong determination of personal directory
## Version of OpenTTD
Tested on trunk 90a4737bf6f082a3620de1649b5dede4af5cb89f
Some calls to the `DEBUG` macro have been added to track the problem down & produce the output pastes that follow.

## Expected result
Normal game runs fine after OpenGRF baseset has been downloaded from ingame, and thus resides in `~/.openttd/content_download/baseset/`
The regression test execution should find it aswell.

A patched regression run (one where the custom configuration file `-c ai/regression/regression.cfg` has been removed) shows as follow:
```
dbg: [misc] Scanning for tars
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/content_download/baseset/, recursive = true
dbg: [misc] Found dir in tar: opengfx-0.5.2/
dbg: [misc] Found file in tar: opengfx-0.5.2/license.txt (17987 bytes, 1024 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/opengfx.obg (9998 bytes, 19968 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/ogfxt_toyland.grf (622079 bytes, 30720 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/ogfxe_extra.grf (911491 bytes, 653312 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/ogfxc_arctic.grf (287331 bytes, 1565696 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/ogfx1_base.grf (2698330 bytes, 1853952 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/ogfxi_logos.grf (253846 bytes, 4553216 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/ogfxh_tropical.grf (472020 bytes, 4807680 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/readme.txt (14584 bytes, 5280256 offset)
dbg: [misc] Found file in tar: opengfx-0.5.2/changelog.txt (24910 bytes, 5295616 offset)
dbg: [misc] Found tar '/home/<user>/.openttd/content_download/baseset/OpenGFX-0.5.2.tar' with 10 new files
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/content_download/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/content_download/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/content_download/data/, recursive = true
dbg: [misc] Scan complete, found 1 files
```
No error is reported.

## Actual result
Regression test execution fails to find the OpenGRF archive, which yields an error.
Moreover, debug output shows the following:
```
dbg: [misc] Scanning for tars
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = ai/regression/content_download/baseset/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = ai/regression/content_download/gm/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = ai/regression/content_download/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/.openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /home/<user>/source/bin/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = /usr/local/share/games/openttd/data/, recursive = true
dbg: [misc] /home/<user>/source/src/fileio.cpp:1363 ScanPath extension = .tar, path = ai/regression/content_download/data/, recursive = true
dbg: [misc] Scan complete, found 0 files
```

## Steps to reproduce
Execute `make regression` on Linux
To showcase the run *not* failing to find the baseset, edit `bin/ai/regression/run.sh` to remove the `-c ai/regression/regression.cfg` argument to the call to `openttd`.

## Quick analysis
The following lines of `fileio.cpp` determine the `config_dir` variable content:
```
1184     if (_config_file != NULL) {
1185         config_dir = stredup(_config_file);
```
Then, the following line replicate `config_dir` in `_personal_dir`, which looks problematic:
```
1237     {
1238         _personal_dir = config_dir;
1239     }
```

### PR DESCRIPTION
**Pull request by Berbe**
`_personal_dir` gets replaced by `config_dir` outside of the XDG triggers if a configuration file is specified.
This makes directories path computation fail to find content sitting in normal game arborescence, such as the autoload directory.

This prosposal aims at preventing this behaviour, by not touching the arborescence when a configuration file is specified.
Discovered & loaded files should thus be more consistent across executions.

Any drawback?

### CODE REVIEW

**Comment by Berbe** [comment]
At least, this makes regression tests pass without patching the script, now that OpenGFX (downloaded from in-game) is found:
```
Running ai/regression/tst_regression... passed!
Running ai/regression/tst_stationlist... passed!
```

**Comment by glx22** [comment]
Hmm I think config_dir should be an extra search dir appended before personal_dir. Would make more sense and keep the current behaviour too. Quote from readme:

> Savegames will be relative to the config file only if there is no save/ directory in paths with higher priority than the config file path, but autosaves and screenshots will always be relative to the config file. Unless the configuration file is in $XDG_CONFIG_HOME/openttd, then all other files will be saved under $XDG_DATA_HOME/openttd.

**Comment by Berbe** [comment]
⚠[EDIT] Mislead comment, keeping it only for reference⚠

----------

> Current proposal breaks past behaviour of OpenTTD, as the location of the configuration file supplied with `-c` implies a configuration *directory* when resources & directories are created/searched for.
> Worse, the configuration directory has precedence for every resource that are now not discovered/loaded anymore.
> 
> That is IMHO bad design, as documented directories shall be the only place the program shall ever write to, unless explicitely told otherwise. Widening the scope of the switch is not to be expected
> With that idea in mind, `-c` should only provide the configuration file and not imply anything else, the working directory structure remaining the same.
> This fix actually enforces that.
> 
> Now, it is not a satisfactory decision, as it breaks the backwards compitibility.
> One idea is to transform the meaning of the `-c` switch from 'Loading a configuration file' to 'Changing working directory': This way, creation/search of files in a specified *directory* would be explicit, thus naturally expected.
> In this working directory, the configuration file would naturally be seeked for and have higher precedence than other locations.
> While the overall behaviour shall be the same, the semantics at hand imply an implementation which might differ quite a bit. I have no idea yet.
> 
> As a fallback, the new `-c` switch, normally accepting directories, could still be accepting the configuration file, implying the working dir around it, exactly as it is doing in the trunk version.

**Comment by Berbe** [comment]
TL; DR
I do not think my PR breaks how the game loads save files in anyway.
From what I witnessed, it merely changes/fixes the path for auto-downloaded content, which was searched in the same location as the supplied game configuration file.

-----

Scratch my long previous comment.

It seems the intel I got about nice precendence of directories over others is simply not how the game is working at the moment.
As an example, savegames are loaded following that logic: https://github.com/OpenTTD/OpenTTD/blob/c0c8fb25fb84d498207713f89720e38a98f16435/src/saveload/saveload.cpp#L2771
The paths are built following a combinations of values stored in those structures: https://github.com/OpenTTD/OpenTTD/blob/c0c8fb25fb84d498207713f89720e38a98f16435/src/fileio_type.h#L110

As I pointed out in my issue (and fixed here), the same `Searchpath` is not necessarily the same used for building paths to each `Subdirectory`. Combinations are context-dependent, and the logic hard to follow.

If you want to question potential side effects, it is best to take them one by one and either follow their logic into the code, statically, or just dynamically compare trunk & this version builds.

**Comment by Berbe** [comment]
Ping?

**Comment by Berbe** [comment]
Ping?

**Comment by LordAro** [comment]
I'm going to close this one, as there's been no consensus on the correct solution, other than this is likely just a hack on top of a (much older) hack. Feel free to discuss in #7311
